### ISSUE TITLE
[Feature Request] Multiplier-Based Image Selection Over Repeat Counts - Improved Subject Balancing

### ISSUE THREAD

**Issue by a-l-e-x-d-s-9** [kohya-ss/sd-scripts#826]
I'm used to [EveryDread2](https://github.com/victorchall/EveryDream2trainer) trainer, which gives users the option to select a portion of images from a folder using a multiplier instead of a repeat count. For instance, by applying a 0.5 multiplier to a folder, the trainer will randomly select half of the images for each epoch. This system isn't limited to fractions; users can set multipliers such as 1.6 or any other positive value. On the other hand, using a negative multiplier will exclude the folder.

This multiplier approach retains the repeat functionality users are accustomed to while offering a significant improvement in balancing training across different subjects without having to delete any images.

To illustrate, consider three subjects: A with 1000 images, B with 200 images, and C with 100 images. If the goal is to train on 25 images per epoch, we can set multipliers of 0.025 for A, 0.125 for B, and 0.25 for C. This eliminates the need to delete images or train on all 1000 images from subject A. This method effectively balances subjects, especially when the number of images in their datasets varies. As the number and diversity of subjects increase, this approach provides better control over training, avoiding the need to discard images.

I believe that adding this feature to Kohya is essential.

**Comment by a-l-e-x-d-s-9** [comment]
@kohya-ss 
Thank you for your response.
In ED2 the feature is pseudo random and based on seed used for training.
I think it's good enough, and there is no need to force full rotation of images or sync between DataLoaders.
I'll be happy to implement this feature myself, but I couldn't figure out how and when exactly the system is handling repeats - to switch them to portion of files in a folder.
If you can point me to specific file or functions, I would appreciate it.

**Comment by a-l-e-x-d-s-9** [comment]
@kohya-ss There is a small misunderstanding that I want to clear up. 
The multiplier is an extension of the current repeat method; it doesn't change how the system handles full repeats, and users will be able to use it in the same way they do now.
For example, with repeats of 2, the multiplier will be 2, and the system will add all images inside this folder for training exactly twice. There is no randomness in the integer part of the multiplier. 
The difference occurs only in the fractional part of the multiplier. So, for a multiplier of 2.5 - the integer 2 will add all images from the folder twice, and 0.5 will add only 50% additional images from the folder. 
The multiplier won't change system behavior or user experience for users who use repeats. But it will allow for much better subject balancing between multiple subjects in a way that's impossible now.

**Comment by a-l-e-x-d-s-9** [comment]
@kohya-ss 
I added a few examples below with a python code.
The fraction part rotation is handled by a probabilistic selection of images every epoch, there is no need to enforce deterministic rotation.
I want to explain how I used this feature in ED2 with 25 subjects, each subject had a different number of images, between 50 and 300 images for each subject. I calculated (with a script, but it can be done manually) a multiplier for each folder, so each subject will use only 25 images per epoch. That way no images would be discarded, and they all can participate in the training, also the balance between subjects is always kept. Unfortunately, it's not something that can be done with the current implementation of repeats in kohya-ss/sd-scripts.

Here are the prints from the code:
```log
- Example of two consecutive epochs with the same folder, with multiplier 2.5, epochs 1, 2:
**************************************************
Multiplier: 2.5, sources: ['A1', 'B1', 'C1', 'D1']
Total images added due to integer part: 8
Additional images added due to fractional part: 2
Total images for training: 10
Images for training: ['A1', 'B1', 'C1', 'D1', 'A1', 'B1', 'C1', 'D1', 'D1', 'C1']
**************************************************
Multiplier: 2.5, sources: ['A1', 'B1', 'C1', 'D1']
Total images added due to integer part: 8
Additional images added due to fractional part: 2
Total images for training: 10
Images for training: ['A1', 'B1', 'C1', 'D1', 'A1', 'B1', 'C1', 'D1', 'D1', 'B1']
- Different folder, epoch 3:
**************************************************
Multiplier: 1.33, sources: ['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2']
Total images added due to integer part: 8
Additional images added due to fractional part: 3
Total images for training: 11
Images for training: ['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2', 'E2', 'C2', 'H2']
- Another folder with fractions only, epoch 5:
**************************************************
Multiplier: 0.5, sources: ['A3', 'B3', 'C3', 'D3', 'E3', 'F3']
Total images added due to integer part: 0
Additional images added due to fractional part: 3
Total images for training: 3
Images for training: ['B3', 'A3', 'E3']
- Another folder with fractions only, epoch 10:
**************************************************
Multiplier: 0.33, sources: ['A4', 'B4', 'C4', 'D4', 'E4', 'F4', 'G4', 'H4']
Total images added due to integer part: 0
Additional images added due to fractional part: 3
Total images for training: 3
Images for training: ['C4', 'H4', 'G4']

```

Code:
```python
import random


def choose_images_for_training_for_folder(files_list, multiplier, main_seed, epochs):
    # Create an isolated instance of random
    local_random = IsolatedRandom(main_seed + epochs)

    images_for_training = []

    # Split the multiplier into integer and fractional parts
    int_multiplier = int(multiplier)
    frac_multiplier = multiplier - int_multiplier

    # Add all images for the integer part of the multiplier
    for _ in range(int_multiplier):
        images_for_training.extend(files_list)

    # Randomly select a subset of images for the fractional part of the multiplier
    num_to_select = int(len(files_list) * frac_multiplier)
    additional_files = local_random.sample(files_list, num_to_select)

    # Probabilistically choose an extra image if the fractional multiplier results in a non-integer number of images
    if len(files_list) * frac_multiplier > num_to_select:
        additional_file = local_random.choice([file for file in files_list if file not in additional_files])
        additional_files.append(additional_file)

    images_for_training.extend(additional_files)

    # Print the results
    print("*" * 50)
    print(f"Multiplier: {multiplier}, sources: {files_list}")
    print(f"Total images added due to integer part: {len(files_list) * int_multiplier}")
    print(f"Additional images added due to fractional part: {len(additional_files)}")
    print(f"Total images for training: {len(images_for_training)}")
    print("Images for training:", images_for_training)

    return images_for_training





class IsolatedRandom:
    def __init__(self, seed=None):
        self._random = random.Random(seed)

    def seed(self, seed=None):
        self._random.seed(seed)

    def sample(self, population, k):
        return self._random.sample(population, k)

    def random(self):
        return self._random.random()

    def choice(self, seq):
        return self._random.choice(seq)


main_seed = 10

# Test with given examples and fixed seed and epoch
print("- Example of two consecutive epochs with the same folder, with multiplier 2.5, epochs 1, 2:")
choose_images_for_training_for_folder(["A1", "B1", "C1", "D1"], 2.5, main_seed, 1)
choose_images_for_training_for_folder(["A1", "B1", "C1", "D1"], 2.5, main_seed, 2)

print("- Different folder, epoch 3:")
choose_images_for_training_for_folder(["A2", "B2", "C2", "D2", "E2", "F2", "G2", "H2"], 1.33, main_seed, 3)

print("- Another folder with fractions only, epoch 5:")
choose_images_for_training_for_folder(["A3", "B3", "C3", "D3", "E3", "F3"], 0.5, main_seed, 5)

print("- Another folder with fractions only, epoch 10:")
choose_images_for_training_for_folder(["A4", "B4", "C4", "D4", "E4", "F4", "G4", "H4"], 0.33, main_seed, 10)
```

**Comment by a-l-e-x-d-s-9** [comment]
The chance that any particular image would not be trained at all with a multiplier as a fraction(only) is: (1-multiplier)^epochs.
It can happen in certain cases, but it doesn't really matter, the important part is that balancing between subjects is enforced, and all subjects utilize most of their images, and the user can achieve uniform training between subjects easily.
Using current repeats is not a very effective solution for subject balancing, because of a couple of reasons:
1. The pace of the learning per subject isn't controlled with a huge pool of repeats of different subjects pooled together. The distribution of subjects can be very uneven and chronic episodes of overtraining for certain subjects, undertraining for others, forgetting of a subject, and other edge cases, can occur when using a huge pool of unbalanced subjects. With a multiplier, it's easy to use 25 images per subject per epoch, and get even training for all subjects.
2. I gave an example of a model I actually trained with 25 subjects ([link to model](https://civitai.com/models/113337/divas)), also I've trained another model with 50 subjects  ([link to model](https://civitai.com/models/129461/game-of-thrones)). In both of them, I used the multiplier to balance between subjects in a very precise way, which allowed me to use a subject with arbitrary amounts of images without the need to discard any image. In my second model, each subject had between 50 and 800 images, and with 50 subjects, I don't think there is an easy way to calculate the repeated subject, and a user still needs to go over all subjects and remove images to make it work with integer repeats. Using a multiplier, the calculation is very easy: multiplier=desired_images_per_epoch/images_per_subject.

**Comment by chillpixelfun** [comment]
@kohya-ss are subfolders with repeats supported?

@a-l-e-x-d-s-9 I like the concept. Would a solution like this work for you? It accomplishes the same outcome by reversing the multiplier using subfolder repeats.

```
1_person/
├─ 2_person_body/
│  ├─ B.png
├─ 2_person_behind/
│  ├─ C.png
├─ 3_person_face/
│  ├─ D.png
│  ├─ E.png
├─ A.png
```

```
repeat count 1, sources: ['A']
repeat count 2, sources: ['B', 'C']
repeat count 3, sources: ['D', 'E']
```

**Comment by a-l-e-x-d-s-9** [comment]
@chillpixelfun
You can achieve nested multiplier behavior by flattening the folder's hierarchy and multiplying the nested repeats - it's identical to the current system.
A repeat of subfolders doesn't address the issues I raised and it doesn't make it easier to balance between subjects.

**Comment by chillpixelfun** [comment]
@a-l-e-x-d-s-9 what issues did you raise that are not addressed? and how does it not make it easier to balance between subjects? The subjects are explicitly balanced using a base folder repeat of 1 and putting the desired multiplier on the subfolders. The subfolder repeat is the multiplier.

**Comment by a-l-e-x-d-s-9** [comment]
@kohya-ss 

Yes, with the introduction of random images per epoch, the bucketing per epoch will require changes.
My idea of implementation consists of two stages:

1. The Initialization Stage: This is similar to the current bucketing. However, it only records to which bucket each image belongs and notes the multiplier of the image.
2. At the beginning of every epoch, images are added to buckets for training. Those with an integer multiplier part above 0 will always be added (according to number). The fractional part will be added based on a probabilistic decision.

To optimize the second step, we might use two lists: permanent and probabilistic. The permanent list is reserved for images with an integer part of the multiplier above 0. For example, images with a multiplier of 2.5 will always be trained twice per epoch, so we can keep them in the permanent list for each epoch's training. The probabilistic list will only contain images with a multiplier that has a fraction. So, if we only have images in the permanent list and none in the probabilistic list, the system will operate pretty much as it does now, with no runtime penalty for the user. However, if the user opts to use multipliers, the probabilistic list's images will be added to the training set based on the fraction of the multiplier at the start of an epoch, and comparison with a random value, if the random value is less than the multiplier's fraction - image will be added.

As far as I understand, the implementation of [ED2](https://github.com/victorchall/EveryDream2trainer) is doing something similar now, and people are using it with hundreds of thousands of images, so it's a valid solution. Regular users, with a few thousands of images, definitely wouldn't notice any changes in performance.
